src/pcmp.rs

```diff
@@ -491,7 +513,21 @@ fn test_find() {
 /// Load the first 16 bytes of `pat` into a SIMD vector.
 #[inline(always)]
 fn pat128(pat: &[u8]) -> __m128i {
-    unsafe { _mm_loadu_si128(pat.as_ptr() as *const _) }
+    unsafe {
+        mask_load(pat.as_ptr() as *const _, pat.len())
+    }
+}
+
+#[inline(always)]
+unsafe fn mask_load(ptr: *const u8, len: usize) -> __m128i {
+    const REGSZ: usize = mem::size_of::<__m128i>();
+    if len >= REGSZ {
+        return _mm_loadu_si128(ptr as _);
+    }
+
+    let mut data = [0; REGSZ];
```

Comment:
This is actually pretty much how `_mm_loadu_si128` is implemented, except for a more efficient init using `_mm_undefined_si128`: https://doc.rust-lang.org/1.27.2/src/core/up/stdsimd/coresimd/x86/sse2.rs.html#1168-1176

So I think it makes sense to remove branching above and instead just copy code of the intrinsic from that source but replace the length.

Reply:
I agree with this

Reply:
As far as I can tell, these intrinsics are not really special and are implemented in pure Rust to simulate corresponding intrinsics from C. However, looking at the assembly of both tweaked version as suggested and current one I see that even though branching goes away, we end up with always hitting non-inlined call to `memcpy` which might be potentially even worse: https://rust.godbolt.org/z/IkxEW8

Reply:
Although I guess most "needles" hit that branch anyway as they are usually smaller than 16 bytes, so might be still more effective to avoid branching.